walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,89 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
-}
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        int TIME_WINDOW_IN_MINUTES = 60 * 24;
+        Integer[] minutesNoOptAttendees = new Integer[TIME_WINDOW_IN_MINUTES];
+        Integer[] minutesOptAttendees = new Integer[TIME_WINDOW_IN_MINUTES];
+
+        Arrays.fill(minutesNoOptAttendees, 1);
+        Arrays.fill(minutesOptAttendees, 1);
+
+        ArrayList<TimeRange> availableTimeRanges = new ArrayList<>();
+        ArrayList<Event> eventsToSet = new ArrayList<>(events);
+
+        Set<String> attendees = new HashSet<String>(request.getAttendees());
+        Set<String> optionalAttendees = new HashSet<String>(request.getOptionalAttendees());
+
+        for (Event event : eventsToSet) { // N
+            /* Verify that at least one of the request attendees
+            is scheduled for the current event. */
+            Set<String> eventAttendees = event.getAttendees();
+            Boolean isAnyoneBusy = !Collections.disjoint(eventAttendees, attendees); // O + P
+            Boolean isAnyOptionalBusy = !Collections.disjoint(eventAttendees, optionalAttendees); // O + Q
+
+            TimeRange when = event.getWhen();
+            int start = when.start();
+            int end = when.start() + when.duration();
+
+            /* Mark available spaces as busy since there's at
+                least one person from the request that isn't available */
+            if (isAnyoneBusy) {
+                Arrays.fill(minutesNoOptAttendees, start, end, 0); // end - start ops.
+            }
+
+            if (isAnyoneBusy || isAnyOptionalBusy) {
+                Arrays.fill(minutesOptAttendees, start, end, 0); // end - start ops.
+            }
+        }
+
+        ArrayList<TimeRange> timeWindowsNoOptAttendees = getTimeWindows(minutesNoOptAttendees); // 1440
+        ArrayList<TimeRange> timeWindowsOptAttendees = getTimeWindows(minutesOptAttendees); // 1440
+
+        long duration = request.getDuration();
+        availableTimeRanges = getAvailableTimeRanges(timeWindowsOptAttendees, duration);
```

Comment:
You can probably avoid this extra pass by passing the `duration` into `getTimeWindows` and checking `windowSize` is larger than the duration before adding it to the returned `ArrayList<TimeRange>`.

Reply:
Yeah, makes sense. Probably I was trying to achieve some modularity and logic encapsulation since that's healthy, I guess. But I agree, this definitely would reduce time and save lines of code.